### ISSUE TITLE
loss为inf

### ISSUE THREAD

**Issue by mengyiliu22** [PaddlePaddle/Paddle#18412]
模型前期训练的loss为inf，请问这是什么原因呢？

**Comment by mengyiliu22** [comment]
[train.log](https://github.com/PaddlePaddle/Paddle/files/3338997/train.log)

附件为训练的loss信息；
损失函数是nce；
优化算法使用Adam和Adagrad都会出现这样的情况；
学习率设为1e-3，不会出现inf，增大到1e-2会出现；
batch_size为512，尝试过多个数值，都会出现inf;
使用ParallelExecutor，单机和多机都会出现这种情况.
从后面的loss看，不会有inf出现了，对模型的训练似乎不会产生影响，主要想弄清楚inf出现的原因，辛苦帮忙看下

**Comment by wangguibao** [comment]
您可以试着在网络中加一些Print，看看到底哪些layer产生了inf，然后针对性的再看下是否通过调整超参能改善。如果一直出，可能是计算逻辑有问题

stackoveflow上这个问题有些tip可以参考：
https://stackoverflow.com/questions/48600374/cross-entropy-loss-suddenly-increases-to-infinity

**Comment by wangguibao** [comment]
如果确定是某些layer的计算逻辑问题，可以提bug修复

**Comment by mengyiliu22** [comment]
> 您可以试着在网络中加一些Print，看看到底哪些layer产生了inf，然后针对性的再看下是否通过调整超参能改善。如果一直出，可能是计算逻辑有问题
> 
> stackoveflow上这个问题有些tip可以参考：
> https://stackoverflow.com/questions/48600374/cross-entropy-loss-suddenly-increases-to-infinity

谢谢老师的解答。
我这边还有一些发现是对exe.run(loss)结果进行打印，发现inf是来源于某个计算节点，别的计算节点数据正常，例如使用了4个cpu，返回4个loss，其中会有inf，也会有正常loss。

**Comment by wangguibao** [comment]
您是分布式训练是么
```
发现inf是来源于某个计算节点，别的计算节点数据正常，例如使用了4个cpu，返回4个loss，其中会有inf，也会有正常loss
```
能具体再描述一下么

**Comment by mengyiliu22** [comment]
> 您是分布式训练是么
> 
> ```
> 发现inf是来源于某个计算节点，别的计算节点数据正常，例如使用了4个cpu，返回4个loss，其中会有inf，也会有正常loss
> ```
> 
> 能具体再描述一下么

分布式和单机训练都会出现；我排查是在单机，使用ParallelExecutor，打印exe.run(loss)的返回值，是个（1，cpu个数）形状的array，array里会出现inf，也会有正常的loss值，所以最后加起来也是inf。。

**Comment by wangguibao** [comment]
经过和 @seiriosPlus  老师一起调试用户提供的可复现代码，基本定位问题如下：
1) 用户网络使用了NCE loss，nce layer的输入input，其数值范围非常大[-100, +100]。
2） 在NCE内部，采样的sample经过一个类似sigmoid的操作 sample_data = (1/1+e^(-sample_data))之后全都变成0或1
3) 在后续的计算cost公式中，cost = -log(o/(o+b)) = Infinity (这个公式里o = 第2)步的sample_data)

初步判断原因是，
NCE loss输入一般直接作为原始的word embedding作为输入，其分量值[-1, +1]，进入NCE后不会有问题；而用户的网络是将原始embedding又经过了2个带relu激活函数的embedding，数值范围变的很大。

已经跟用户解释上述原因，待用户进一步反馈。

**Comment by wangguibao** [comment]
线下和用户沟通，反馈同样的网络在v2上是可以训练没问题的。我建议用户再diff一下网络在v2和fluid上实现有哪些细节上不同，或者可以提需求让paddle团队支持网络打平。

用户说自己先调一下超参什么的

**Comment by wangguibao** [comment]
issue已超过一星期，用户未再反馈，先关闭。如有需要可随时重新打开
